[MXNET-1358]Fit api tutorial

roywei opened the pull request:
## Description ##

1. bring back https://github.com/apache/incubator-mxnet/pull/14462
main contributor @piyushghai 
I have updated the tutorial to the latest fit/estimator api
2. update website API generation( add to API index, fixed imports for automated API doc generation)
3. fixed checkpoint reuqires hybridization


@ThomasDelteil @thomelane @aaronmarkham  

## Preview
* http://mxnet-ci-doc.s3-accelerate.dualstack.amazonaws.com/PR-15353/8/tutorials/gluon/fit_api_tutorial.html

* http://mxnet-ci-doc.s3-accelerate.dualstack.amazonaws.com/PR-15353/8/api/python/gluon/contrib.html?highlight=estimator#module-mxnet.gluon.contrib.estimator

aaronmarkham commented on docs/tutorials/gluon/fit_api_tutorial.md:
```diff
@@ -0,0 +1,265 @@
+                                       save_best=True)  # Save the best model in terms of
+# Let's instantiate another handler which we defined above 
+loss_record_handler = LossRecordHandler()
+# Magic line
```
Lol ok

aaronmarkham commented on docs/tutorials/gluon/fit_api_tutorial.md:
````diff
@@ -0,0 +1,265 @@
+    Epoch 1, loss 0.5741 <!--notebook-skip-line-->
+    Epoch 2, loss 0.3229 <!--notebook-skip-line-->
+
+You can load the saved model, by using ```load_parameters``` API in Gluon. For more details refer to the [Loading model parameters from file tutorial](http://mxnet.incubator.apache.org/versions/master/tutorials/gluon/save_load_params.html#saving-model-parameters-to-file)
````
```suggestion
You can load the saved model, by using the `load_parameters` API in Gluon. For more details refer to the [Loading model parameters from file tutorial](save_load_params.html#saving-model-parameters-to-file)
```

aaronmarkham commented on docs/tutorials/gluon/fit_api_tutorial.md:
````diff
@@ -0,0 +1,265 @@
+
+## Summary
+
+In this tutorial, we learnt how to use ```Gluon Fit APIs``` for training a deep learning model and also saw an option to customize it with the use of Event Handlers.
````
```suggestion
In this tutorial, you learned how to use Gluon Fit APIs for training a deep learning model and also saw an option to customize it with the use of `Event Handlers`.
```
🇺🇸

aaronmarkham commented on docs/tutorials/gluon/fit_api_tutorial.md:
````diff
@@ -0,0 +1,265 @@
+## Summary
+
+In this tutorial, we learnt how to use ```Gluon Fit APIs``` for training a deep learning model and also saw an option to customize it with the use of Event Handlers.
+For more references and advanced usage details can be found in the [documentation](http://mxnet.apache.org/api/python/gluon/gluon.html).
````
```suggestion
More references and advanced usage details can be found in the [Gluon API documentation](http://mxnet.apache.org/api/python/gluon/gluon.html).
```

aaronmarkham commented on docs/tutorials/gluon/fit_api_tutorial.md:
```diff
@@ -0,0 +1,265 @@
+
+## Next Steps 
+
+- To learn more about deep learning with MXNet Gluon, see [Deep Learning - The Straight Dope](https://gluon.mxnet.io)
```
```suggestion
- To learn more about deep learning with MXNeT, see [Dive Into Deep Learning](http://gluon.io)
```
D2L > TSD and D2L doesn't have SSL.

aaronmarkham commented on docs/tutorials/gluon/fit_api_tutorial.md:
```diff
@@ -0,0 +1,265 @@
+## Next Steps 
+
+- To learn more about deep learning with MXNet Gluon, see [Deep Learning - The Straight Dope](https://gluon.mxnet.io)
+- For more hands on learning about deep learning, check out [Dive into Deep Learning](https://d2l.ai)
```
```suggestion

```

anirudhacharya commented:
@mxnet-label-bot add [pr-awaiting-review]

roywei commented on docs/tutorials/gluon/fit_api_tutorial.md:
```diff
@@ -0,0 +1,265 @@
+
+To complete this tutorial, you will need:
+
+- [MXNet](https://mxnet.incubator.apache.org/install/#overview) (The version of MXNet will be >= 1.5.0)
```
if build from source it's already 1.5.0 on master, once 1.5.0 release, master will be 1.6.0. Added explaination

roywei commented on docs/tutorials/gluon/fit_api_tutorial.md:
```diff
@@ -0,0 +1,265 @@
+
+In this tutorial, we will see how to use the [Gluon Fit API](https://cwiki.apache.org/confluence/display/MXNET/Gluon+Fit+API+-+Tech+Design) which is the easiest way to train deep learning models using the [Gluon API](http://mxnet.incubator.apache.org/versions/master/gluon/index.html) in Apache MXNet. 
+
+With the Fit API, you can train a deep learning model with miminal amount of code. Just specify the network, loss function and the data you want to train on. You don't need to worry about the boiler plate code to loop through the dataset in batches(often called as 'training loop'). Advanced users can still do this for bespoke training loops, but most use cases will be covered by the Fit API.
```
changed to many

roywei commented on docs/tutorials/gluon/fit_api_tutorial.md:
```diff
@@ -0,0 +1,265 @@
+
+gpu_count = mx.context.num_gpus()
+ctx = [mx.gpu(i) for i in range(gpu_count)] if gpu_count > 0 else mx.cpu()
+mx.random.seed(7) # Set a fixed seed
```
removed it, if a tutorial is been tested against accuracy (e.g. assert acc > 0.8 after 5 epochs, fixed seed may be added for less flakiness). here we are not checking acc, so removed it

roywei commented on docs/tutorials/gluon/fit_api_tutorial.md:
````diff
@@ -0,0 +1,265 @@
+resnet_18_v1.initialize(init = mx.init.Xavier(), ctx=ctx)
+```
+
+We will be using ```SoftmaxCrossEntropyLoss``` as the loss function since this is a multi-class classification problem. We will be using ```sgd``` (Stochastic Gradient Descent) as the optimizer. You can experiment with a different optimizer as well. 
````
done

roywei commented:
@aaronmarkham I have addressed your comments, could you take another look? thanks!
Another problem is the estimator API doc is not showing up on the website, not sure if [this](https://github.com/apache/incubator-mxnet/pull/15353/files#diff-0e3f58347a224fc8184299f9c154cda1R29) will fix it

abhinavs95 commented on docs/tutorials/gluon/fit_api_tutorial.md:
```diff
@@ -0,0 +1,263 @@
+
+To complete this tutorial, you will need:
+
+- [MXNet](https://mxnet.incubator.apache.org/install/#overview) (The version of MXNet will be >= 1.5.0, you can use `pip install mxnet --pre` to get the latest pip package or build from source with master, refer to [MXNet installation](http://mxnet.incubator.apache.org/versions/master/install/index.html?platform=Linux&language=Python&processor=CPU)
```
We can update this now since 1.5 is out?

abhinavs95 commented on docs/tutorials/gluon/fit_api_tutorial.md:
```diff
@@ -0,0 +1,263 @@
+[Fashion-MNIST](https://research.zalando.com/welcome/mission/research-projects/fashion-mnist/) dataset consists of fashion items divided into ten categories: t-shirt/top, trouser, pullover, dress, coat, sandal, shirt, sneaker, bag and ankle boot. 
+
+- It has 60,000 grayscale images of size 28 * 28 for training.  
+- It has 10,000 grayscale images os size 28 * 28 for testing/validation. 
```
typo: os->of

eric-haibin-lin commented:
any update on this ? mxnet 1.5 is released but no doc for the feature.

roywei commented:
I have addressed commend @aaronmarkham @abhinavs95 could you take another look? Thanks

roywei commented:
I have verified the documentation is successfully generated, tutorial is passing, nightly test are passing and unit test should be already checked by CI

ping @aaronmarkham for review on the docs. Thanks!

![estimator](https://user-images.githubusercontent.com/8022184/62243117-7bca5c00-b391-11e9-9a11-a236f4a70f3c.png)
![checkpoint](https://user-images.githubusercontent.com/8022184/62243124-7ec54c80-b391-11e9-8acd-45bfca464e66.png)

roywei commented:
Update website API docs can be found at: http://mxnet-ci-doc.s3-accelerate.dualstack.amazonaws.com/PR-15353/10/api/python/gluon/contrib.html

Tutorial: http://mxnet-ci-doc.s3-accelerate.dualstack.amazonaws.com/PR-15353/10/tutorials/gluon/fit_api_tutorial.html
